lib/pat.c

```diff
@@ -1119,11 +1119,15 @@ key_put(grn_ctx *ctx, grn_pat *pat, const uint8_t *key, uint32_t len)
   }
 
   /* Check if the key to be added can fit in the current segment. */
-  uint32_t end_segment = (res + len - 1) >> W_OF_KEY_IN_A_SEGMENT;
+  uint64_t end_segment = (res + len - 1) >> W_OF_KEY_IN_A_SEGMENT;
   if (res >> W_OF_KEY_IN_A_SEGMENT != end_segment) {
     /* If it does not fit, update `curr_key`(`res`) to add it to the next
      * segment. */
-    res = pat->header->curr_key = end_segment << W_OF_KEY_IN_A_SEGMENT;
+    if (pat_is_key_large(pat)) {
+      res = pat->header->curr_key_large = end_segment << W_OF_KEY_IN_A_SEGMENT;
+    } else {
+      res = pat->header->curr_key = end_segment << W_OF_KEY_IN_A_SEGMENT;
+    }
```

Comment:
Maybe we could change pat_update_curr_key() and use that.

https://github.com/groonga/groonga/blob/3c9da630de8c49a8306f74755f53ebd6cb179391/lib/pat.c#L5964-L5968